File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    q = deque()
    q.append((0, 0))  
    directions = [(-1,0), (1,0), (0,-1), (0,1)]  

    while q:
        y, x = q.popleft()

        for dy, dx in directions:
            ny, nx = y + dy, x + dx

           
            if 0 <= ny < n and 0 <= nx < m and maps[ny][nx] == 1:
                
                maps[ny][nx] = maps[y][x] + 1
                q.append((ny, nx))

   
    if maps[n-1][m-1] == 1:
        return -1  
    else:
        return maps[n-1][m-1]  

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>

using namespace std;

int solution(vector<vector<int> > maps)
{   int n = maps.size();
    int m = maps[0].size();
    vector<vector<bool>> visited(n, vector<bool>(m, false));
 
    queue<pair<int,int>> q ;
    q.push({0,0});
    visited[0][0] = true;
 
    int dy[4] = {-1,1,0,0};
    int dx[4] = {0,0,-1,1};
    while (!q.empty()){
        int ey = q.front().first;
        int ex = q.front().second;
        q.pop();
        
        for (int i=0; i<4; i++){
            int ny = ey + dy[i];
            int nx = ex + dx[i];
            
            if (ny>=0 && ny < n && nx >=0 && nx<m){
                if (maps[ny][nx] ==1 && visited[ny][nx] ==0) {
                    visited[ny][nx] = true;
                    maps[ny][nx] = maps[ey][ex] + 1;
                    q.push({ny,nx});
                }
            }
            }
        }
 
      if (maps[n-1][m-1] == 1) 
          return -1;
      else 
          return maps[n-1][m-1];
      
    }
    